File: sdk/java/src/main/java/com/anyvali/schemas/EnumSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Schema that accepts one of a fixed set of values.
 */
public class EnumSchema extends Schema<Object> {
    private final List<Object> values;

    public EnumSchema(List<Object> values) {
        this.values = new ArrayList<>(values);
    }

    private EnumSchema(EnumSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.values = new ArrayList<>(other.values);
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        // Check for type-strict match
        for (Object v : values) {
            if (v == null && input == null) return input;
            if (v != null && input != null) {
                // Allow Number cross-type matching (Long/Integer/etc)
                if (v instanceof Number && input instanceof Number
                        && !(v instanceof Boolean) && !(input instanceof Boolean)) {
                    if (((Number) v).longValue() == ((Number) input).longValue()
                            && ((Number) v).doubleValue() == ((Number) input).doubleValue()) {
                        return input;
                    }
                } else if (v.getClass().equals(input.getClass()) && v.equals(input)) {
                    return input;
                }
            }
        }

        String valuesStr = values.stream()
                .map(String::valueOf)
                .collect(Collectors.joining(","));
        ctx.addIssue(IssueCodes.INVALID_TYPE,
                "Expected one of enum(" + valuesStr + "), received " + input,
                "enum(" + valuesStr + ")", String.valueOf(input));
        return null;
    }

    @Override
    protected Map<String, Object> toNode() {
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "enum");
        node.put("values", new ArrayList<>(values));
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema<Object> copy() {
        return new EnumSchema(this);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/EnumSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class EnumSchema(
    val values: List<Any?>
) : Schema() {
    override val kind: String = "enum"

    override fun validateValue(input: Any?, ctx: ValidationContext): List<ValidationIssue> {
        val matched = values.any { v -> valuesEqual(input, v) }
        if (!matched) {
            val enumStr = values.joinToString(",") { it?.toString() ?: "null" }
            return listOf(
                ValidationIssue(
                    code = IssueCodes.INVALID_TYPE,
                    path = ctx.path,
                    expected = "enum($enumStr)",
                    received = input?.toString() ?: "null"
                )
            )
        }
        return emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("enum"))
        put("values", JsonArray(values.map { v ->
            when (v) {
                null -> JsonNull
                is String -> JsonPrimitive(v)
                is Boolean -> JsonPrimitive(v)
                is Long -> JsonPrimitive(v)
                is Int -> JsonPrimitive(v)
                is Double -> JsonPrimitive(v)
                is Number -> JsonPrimitive(v.toDouble())
                else -> JsonPrimitive(v.toString())
            }
        }))
    }

    private fun valuesEqual(a: Any?, b: Any?): Boolean {
        if (a == null && b == null) return true
        if (a == null || b == null) return false
        if (a is Boolean && b is Boolean) return a == b
        if (a is Boolean || b is Boolean) return false
        if (a is String && b is String) return a == b
        if (a is String || b is String) return false
        if (a is Number && b is Number) {
            return Schema.toDouble(a) == Schema.toDouble(b)
        }
        return a == b
    }
}
